tests/unit/test_systemd.py

```diff
@@ -9,38 +10,41 @@
 
 
 def with_mock_subp(func):
-    """Set up a Popen mock.
+    """Set up a `subprocess.run(...)` mock.
 
     Any function that uses this decorator should take a function as an argument. When
-    called with a series of return codes, that function will mock out subprocess.Popen,
+    called with a series of return codes, that function will mock out `subprocess.run(...)`,
     and set it to return those error codes, in order.
 
-    The function returns the mock Popen object, so that routines such as
-    assert_called_with can be called upon it, along with the return from
-    systemd._popen_kwargs, for convenience when composing call objects.
-
+    The function returns the mock `subprocess.run(...)` object, so that routines such as
+    assert_called_with can be called upon it.
     """
 
     @patch("charms.operator_libs_linux.v1.systemd.subprocess")
     def make_mocks_and_run(cls, mock_subp):
         def make_mock_popen(returncodes: List[int], lines: List[str] = None, stdout: str = None):
             lines = lines if lines is not None else ("", "")
 
-            mock_subp.PIPE = mock_subp.STDOUT = stdout or ""
+            mock_subp.PIPE = mock_subp.STDOUT = stdout or subprocess.PIPE
 
             side_effects = []
 
             for code in returncodes:
                 mock_proc = MagicMock()
-                mock_proc.wait.return_value = None
-                mock_proc.stdout.readline.side_effect = lines
                 mock_proc.returncode = code
+                mock_proc.stdout = lines
                 side_effects.append(mock_proc)
 
-            mock_popen = mock_subp.Popen
-            mock_popen.side_effect = tuple(side_effects)
+            mock_run = mock_subp.run
+            mock_run.side_effect = tuple(side_effects)
 
-            return mock_popen, systemd._popen_kwargs()
+            return mock_run, {
+                "stdout": subprocess.PIPE,
+                "stderr": subprocess.PIPE,
+                "text": True,
+                "bufsize": 1,
+                "encoding": "utf-8",
+            }
```

Comment:
Appreciate this, thanks! Good to avoid `_private` funcs in tests, and to spell things out separately, otherwise you're just reusing (potentially) buggy data in the test.